Add `--recompiles` to print recompile stats in benchmarks

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #94710

Ex, static:
```
python  benchmarks/dynamo/benchmarks.py --backend aot_eager --training  --accuracy --timing --only hf_Reformer --recompiles
```
```
[["'forward' (modeling_reformer.py:1467)", 1], ["'<graph break in _init_attention_seed>' (modeling_reformer.py:1443)", 11], ["'<graph break in forward>' (modeling_reformer.py:1484)", 5], ["'<graph break in _init_feed_forward_seed>' (modeling_reformer.py:1460)", 11], ["'<graph break in forward>' (modeling_reformer.py:1509)", 5]]
```

Ex, dynamic:
```
TORCHDYNAMO_DYNAMIC_SHAPES=1  AOT_DYNAMIC_SHAPES=1 python  benchmarks/dynamo/benchmarks.py --backend aot_eager --training  --accuracy --timing --only hf_Reformer --recompiles
```
```
[["'forward' (modeling_reformer.py:1467)", 1], ["'<graph break in _init_attention_seed>' (modeling_reformer.py:1443)", 11], ["'<graph break in forward>' (modeling_reformer.py:1484)", 5], ["'<graph break in _init_feed_forward_seed>' (modeling_reformer.py:1460)", 11], ["'<graph break in forward>' (modeling_reformer.py:1509)", 5]]
```
(same) 

cc @mlazos @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

voznesenskym committed:
recompile stats

A little extra

Lint

[ghstack-poisoned]

ezyang commented on benchmarks/dynamo/common.py:
```diff
@@ -1703,6 +1710,9 @@ def get_example_inputs(self):
     )
     parser.add_argument("--timing", action="store_true", help="Emits phase timing")
+    parser.add_argument(
+        "--recompiles", action="store_true", help="Emits recompile stats"
```
Pick a different name, this is too close to verb `--recompile` and not at all clear what it does. `--print-recompiles` is an obvious alternate.

ezyang commented:
I don't really understand how I'm supposed to use this information. The example prints have way to much information, and they seem to all mention graph breaks; is this about recompiles or is about graph breaks? I could see a single flat "you have recompiled this many times" counter being useful for tracking regressions, though in the benchmark suite it's not clear to me that you will trigger this? I also wonder if a flat stat is that useful, because a recompile on a big block is disastrous but a recompile on a small block might just mean you have a bunch of graph breaks.

voznesenskym commented:
> I don't really understand how I'm supposed to use this information. 

 It is very useful to have dump-able info for comparing per-function recompiles. I plan to use it to help guide guard simplification, as well as potential other features I want to explore like an @unguarded annotation.

> The example prints have way to much information, and they seem to all mention graph breaks; 

This is a pervasive source of annoyance for people using dynamo - this is the name of a function rewritten through `resume_execution`

```
code_options["co_name"] = f"<graph break in {code_options['co_name']}>"
            code_options["co_firstlineno"] = lineno
            code_options["co_cellvars"] = tuple()
            code_options["co_freevars"] = freevars
            code_options["co_argcount"] = len(args)
            code_options["co_posonlyargcount"] = 0
            code_options["co_kwonlyargcount"] = 0
```

> is this about recompiles or is about graph breaks? 

Recompiles

> I could see a single flat "you have recompiled this many times" counter being useful for tracking regressions, though in the benchmark suite it's not clear to me that you will trigger this?

Wait, why not? Even on a single iteration of a forward for a model, we could have many recompiles for a given frame, right?

> I also wonder if a flat stat is that useful, because a recompile on a big block is disastrous but a recompile on a small block might just mean you have a bunch of graph breaks.

This isn't for the spreadsheet or anything, I don't really care for a flat stat.

voznesenskym committed:
Update on "Add `--recompiles` to print recompile stats in benchmarks"


Ex, static:
```
python  benchmarks/dynamo/benchmarks.py --backend aot_eager --training  --accuracy --timing --only hf_Reformer --recompiles
```
```
[["'forward' (modeling_reformer.py:1467)", 1], ["'<graph break in _init_attention_seed>' (modeling_reformer.py:1443)", 11], ["'<graph break in forward>' (modeling_reformer.py:1484)", 5], ["'<graph break in _init_feed_forward_seed>' (modeling_reformer.py:1460)", 11], ["'<graph break in forward>' (modeling_reformer.py:1509)", 5]]
```

Ex, dynamic:
```
TORCHDYNAMO_DYNAMIC_SHAPES=1  AOT_DYNAMIC_SHAPES=1 python  benchmarks/dynamo/benchmarks.py --backend aot_eager --training  --accuracy --timing --only hf_Reformer --recompiles
```
```
[["'forward' (modeling_reformer.py:1467)", 1], ["'<graph break in _init_attention_seed>' (modeling_reformer.py:1443)", 11], ["'<graph break in forward>' (modeling_reformer.py:1484)", 5], ["'<graph break in _init_feed_forward_seed>' (modeling_reformer.py:1460)", 11], ["'<graph break in forward>' (modeling_reformer.py:1509)", 5]]
```
(same) 

cc mlazos soumith yanboliang penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

ezyang commented:
> It is very useful to have dump-able info for comparing per-function recompiles. I plan to use it to help guide guard simplification, as well as potential other features I want to explore like an @unguarded annotation.

So, in general, I would like to land things like this even if they're not fully baked, because we're master only and there's value to sharing things around (and annoying to keep having to repatching some local instrumentation that you've found useful.) We can always tweak it as we discover what's useful or not.

However, I need to balance this with not adding a bajillion knobs that need to be maintained by anyone else working on the code in the future, long after the original author has forgotten what they are for. So, my bare minimum requirement here is that *someone else* should know how to use this. I could be that person. But from just looking at the PR, I don't know how to use it, nor is there any instructions on how to usefully use this data elsewhere. Maybe @jbschlosser has a different opinion? In any case, you should explain to at least one of us how to actually use it. Since you are printing machine friendly format, I assume there's some postprocessing to do on it? Checking in some code to scripts/analysis that makes use of it would go a long way towards this.

> This is a pervasive source of annoyance for people using dynamo - this is the name of a function rewritten through resume_execution

Is co_name giving much utility then? Maybe better to only report file name and line no.

> Wait, why not? Even on a single iteration of a forward for a model, we could have many recompiles for a given frame, right?

Well yes, I agree this is useful because it would have identified the Deberta Sequential recompile problem. But in the steady state I expect the number of recompiles to be zero (good thing to put in the CI!)
